proposals/top-level-members.md

```diff
@@ -0,0 +1,216 @@
+
+## Alternatives
+
+- Support `args` keyword in top-level members (just like it can be accessed in top-level statements). But we have `System.Environment.GetCommandLineArgs()`.
+- Allow capturing variables from top-level statements inside non-`static` top-level members.
+  Could be used to refactor a single-file program into multi-file program just by extracting functions to separate files.
+  But it would mean that a method's implementation (top-level statements) can influence what other methods see (which variables are available in top-level members).
+- Allow declaring top-level members outside namespaces as well.
+  - Would introduce ambiguities with top-level statements.
+  - Could be brought to scope via `extern alias`.
+    - To avoid needing to specify those in project files (e.g., so file-based apps also work),
+      there could be a syntax for that like `extern alias Util = Util.dll`.
+    - Or they could be in scope only in the current assembly.
+- Allow declaring top-level statements inside namespaces as well.
+  - Top-level local functions would introduce ambiguities with top-level methods. Wouldn't be a breaking change though, just need to decide which one wins.
+- If we ever allow the `file` modifier on members (methods, fields, etc.), that would be naturally useful for top-level members, too.
+  `file` members would be scoped to the current file.
+  Compare that with `private` members which are scoped to the current _namespace_.
+- Scoping concerns could be resolved with [file-scoped types](https://github.com/dotnet/csharplang/discussions/928) by allowing something like `class MyNamespace.MyClass;`.
+  That wouldn't solve the use-site though, where you'd still need `using static MyNamespace.MyClass;` instead of just `using MyNamespace;` as with this proposal.
+
+- We could have something similar to VB's modules which are mostly like static classes
+  but their members don't need to be qualified with the module name if they are brought to scope via an import:
+  ```vb
+  Imports N
+
+  Namespace N
+      Module M
+          Sub F()
+          End Sub
+      End Module
+  End Namespace
+
+  Class C
+      Sub Main()
+          F()
+      End Sub
+  End Class
+  ```
+
+  F# has something similar, too:
+  ```fs
+  module Utilities =
+    let M() = ()
+
+  open Utilities
+  M()
+  ```
+
+## Open questions
+
+- Which member kinds? Methods, fields, properties, indexers, events, constructors, operators.
+- Accessibility: what should be the default and which modifiers should be allowed?
+- Clustering: currently each namespace per assembly gets its `TopLevel` class.
+- Shape of the synthesized static class (currently `[TopLevel] TopLevel`).
+- Should we simplify the TLS entry point logic? Should it be a breaking change?
+- Should we require the `static` modifier (and keep our doors open if we want to introduce some non-`static` top-level members in the future)?
```

Comment:
fwiw, i think we should require 'static'.  It seems very odd to me that these have 'instance' syntax but aren't referenced/invoked as instancs off of soemthing.  As these are associated with namespaces, and you can't have 'instances' of namespaces, making these look like instance seems weird.  This is esp true as you would be able to write things like `N.SomeNSMethod();`.   That strongly screams "i am static".   

Note: this also follows how we require 'static' on members, even when in a 'static class'.  Just because they must be 'static' doesn't mean we think that 'static' is then not required to say.  

TO me, this makes the declaration/semantics extremely clear, and logically fits in well with other parts of the language (including those that this is directly sitting on top of).